src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java

```diff
@@ -0,0 +1,153 @@
+        return CAPABILITIES_URL;
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
+        mockService.shutdown();
+    }
+
+    protected static final String BASE_REQUEST =
+            "wms?LAYERS=cascadedLayer"
+                    + "&STYLES=&FORMAT="
+                    + MapMLConstants.MAPML_MIME_TYPE
+                    + "&SERVICE=WMS&VERSION=1.3.0"
+                    + "&REQUEST=GetMap"
+                    + "&SRS=EPSG:4326"
+                    + "&BBOX=0,0,1,1"
+                    + "&WIDTH=150"
+                    + "&HEIGHT=150"
+                    + "&format_options="
+                    + MapMLConstants.MAPML_WMS_MIME_TYPE_OPTION
+                    + ":image/png";
+
+    protected XpathEngine xpath;
+
+    protected static WireMockServer mockService;
+
+    @Override
+    protected void registerNamespaces(Map<String, String> namespaces) {
+        namespaces.put("wms", "http://www.opengis.net/wms");
+        namespaces.put("ows", "http://www.opengis.net/ows");
+        namespaces.put("html", "http://www.w3.org/1999/xhtml");
+    }
+
+    @Before
+    public void setup() {
+        xpath = XMLUnit.newXpathEngine();
+    }
+
+    protected void checkCascading(String path, boolean shouldCascade, String rel) throws Exception {
+        Document doc = getMapML(path);
+        // printDocument(doc);
+        String url = xpath.evaluate("//html:map-link[@rel='" + rel + "']/@tref", doc);
+        assertCascading(shouldCascade, url);
+
+        url = xpath.evaluate("//html:map-link[@rel='query']/@tref", doc);
+        assertCascading(shouldCascade, url);
+    }
+
+    protected void assertCascading(boolean shouldCascade, String url) {
+        if (shouldCascade) {
+            assertTrue(
+                    url.startsWith(
+                            "http://localhost:" + mockService.port() + MOCK_SERVER + CONTEXT));
+            assertTrue(url.contains("layers=topp:states"));
+        } else {
+            assertTrue(url.startsWith("http://localhost:8080/geoserver" + CONTEXT));
+            assertTrue(url.contains("layers=cascadedLayer"));
+        }
+    }
+
+    /**
+     * Executes a request using the GET method and returns the result as an MapML document.
+     *
+     * @param path The portion of the request after the context, example:
+     * @return A result of the request parsed into a dom.
+     */
+    protected org.w3c.dom.Document getMapML(final String path) throws Exception {
+        MockHttpServletRequest request = createRequest(path, false);
+        request.addHeader("Accept", "text/mapml");
+        request.setMethod("GET");
+        request.setContent(new byte[] {});
+        String resp = dispatch(request, "UTF-8").getContentAsString();
+        return dom(new ByteArrayInputStream(resp.getBytes()), true);
+    }
+
+    /** For debugging purposes */
+    public static void printDocument(Document doc) throws TransformerException {
```

Comment:
Why not just use the existing "GeoServerSystemTestSupport.print(Document)"  method?
It's already set up to write on System.out when running tests from IDE, but skip prints when running them over a maven build.

Reply:
Ah. I didn't know about that. Nice. Updating it accordingly.